changed function signature and logic to allow for optional height and…

MaxChen102 opened the pull request:
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_

This PR adds optional width and height attributes to every object, allowing users to customize how big the object is.

<details>
<summary>Screenshots of your changes (if applicable)</summary>
<img width="210" height="69" alt="Screenshot 2025-07-10 120053" src="https://github.com/user-attachments/assets/02a44dac-ad6b-47e1-83ae-f2eeeb84d0cb" />

<img width="103" height="257" alt="Screenshot 2025-07-10 120109" src="https://github.com/user-attachments/assets/2a5f4a19-a2b2-41f8-9ab8-cf2fb2d63189" />

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |   X      |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [X] I have performed a self-review of my changes.
    - Check that all changed files included in this pull request are intentional changes.
    - Check that all changes are relevant to the purpose of this pull request, as described above.
- [X] I have added tests for my changes, if applicable.
    - This is **required** for all bug fixes and new features.
- [X] I have updated the project documentation, if applicable.
    - This is **required** for new features.
- [ ] If this is my first contribution, I have added myself to the [list of contributors](https://github.com/david-yz-liu/memory-viz/blob/master/memory-viz/package.json).
- [X] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

- [X] I have verified that the CI tests have passed.
- [X] I have reviewed the test coverage changes reported by Coveralls.
- [X] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

MaxChen102 commented:
Hi @david-yz-liu , I’ve written more tests, please let me know what you think.

david-yz-liu commented on memory-viz/src/memory_model.ts:
```diff
@@ -236,26 +253,47 @@ export class MemoryModel {
+            );
+        }
+
+        let box_width =
```
In general avoid ternary operators that are split across multiple lines. You can make this expression more concise by doing `Math.max(width ?? 0, default_width)`

david-yz-liu commented on memory-viz/src/tests/draw.spec.tsx:
```diff
@@ -530,6 +530,85 @@ describe("draw function", () => {
+            roughjs_config: { options: { seed: 12345 } },
+        });
+        expect(spy).toHaveBeenCalledTimes(3);
+        console.log(spy.mock.calls[0][0]);
```
There should not be any `console.log` statements in the tests

david-yz-liu commented on memory-viz/src/automate.ts:
```diff
@@ -189,20 +189,20 @@ function drawAutomatedOtherItems(
+                        " diagram."
+                );
+            }
+            item.layoutHeight = item.height;
```
Overall, we should not have to use a separate `layoutHeight` attribute.

Instead, if you need to override the user-provided width and height, then overwrite the `height` and `width` fields directly.

Also, I don't think it makes sense to set this attributes for `.blank` objects, since the whole point is they're supposed to be skipped in the diagram if their width/height is not provided.

MaxChen102 commented:
Hi @david-yz-liu , thanks for the feedback. I've removed the attributes `layoutHeight` and `layoutWidth`, and reverted the changes made to `automate.ts`. I made changes to `drawPrimitive()` to shrink the width and height if they are defined so when `drawAll()` is called a second time in `drawAutomatedOtherItems()`, the defined width and height will be adjusted for the inner rectangle. Please let me know what you think.

MaxChen102 commented on docs/docs/02-entity_specification.md:
```diff
@@ -245,17 +261,29 @@ All supported entities include the following core attributes, unless explicitly
+| `value`           | `object`            | Yes      | A mapping of attribute names (as keys) to object IDs (as values). Attribute names may be empty or all-whitespace strings and attribute values may be `null` to render blank boxes, respectively. |
+| `x`, `y`          | `number`            | No       | Optional manual coordinates. Required only if `automation` is disabled.                                                                                                                          |
+| `style`           | `object` or `array` | No       | Custom visual styling.                                                                                                                                                                           |
+| `width`, `height` | `number`            | No       | The height and width for the class.                                                                                                                                                              |
```
Yea it does, just added it!

MaxChen102 commented:
Hi @david-yz-liu , I've made the changes you requested, and added screenshots of the SVG outputs in the PR description.
